### PR TITLE
[PINOT-4926] Introduce a memory manager so that we can switch between Mmap and direct allocation

### PR DESCRIPTION
**Pull request by mcvsubbu**
Introduced the abstract class OffHeapMemoryManager, and the two concrete sub-classes DirectMemoryManager
and MmapMemoryManager.

Re-factored all the classes allocating memory for real-time segments to use the OffHeapMemoryManager abstract
class to allocate memory.

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/DirectMemoryManager.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// Allocates memory using direct allocation
+public class DirectMemoryManager extends OffHeapMemoryManager {
+
+  public DirectMemoryManager(final String segmentName) {
+    super(segmentName);
+  }
+
+  @Override
+  public PinotDataBuffer allocate(long size, String columnName) {
+    return PinotDataBuffer.allocateDirect(size, getSegmentName() + "." + columnName);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/DirectMemoryManager.java]: you need to add this buffer to _buffers list to correctly release memory

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    return _segmentName;
+  }
+
+  /**
+   * Allocate memory
+   * @param size size of memory
+   * @param columnName Name of the column for which memory is being allocated
+   * @return PinotDataBuffer
+   */
+  public abstract  PinotDataBuffer allocate(long size, String columnName);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: See comment on DirectBuffers not adding buffer to _buffers which can cause leak.

It may be better to provide abstract protected allocateInternal method (similar to doClose) that derived classes can implement to provide buffer. This method can add that to _buffers (since _buffers is a member of this class)

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  public abstract void doClose();
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: private

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  /**
+   * Close out this memory manager and release all memory and resources
+   * @throws IOException
+   */
+  public void close() throws IOException {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: Consider logging and swallowing IOException here. Higher layers can hardly take action on that

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+/**
+ * @class OffHeapMemoryManager is an abstract class that implements base functionality to allocate and release
+ * memory that is acquired during realtime segment consumption.
+ *
+ * Realtime consuming segments use memory for dictionary, forward index, and inverted indices. For off-heap
+ * allocation of memory, we instantiate one OffHeapMemoryManager for each segment in the server,
+ *
+ * Closing the OffHeapMemoryManager also releases all the resources allocated by the OffHeapMemoryManager.
+ */
+public abstract class OffHeapMemoryManager implements Closeable {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: I prefer IndexOffHeapMemoryManager given limited scope of memory management of this class

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  protected final List<PinotDataBuffer> _buffers = new LinkedList<>();
+  protected final String _segmentName;
+
+  protected OffHeapMemoryManager(String segmentName) {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: (If I correctly understand that this is Index level memory management) constructor argument should be 'context' or segmentName + indexName. Our Current allocationContext uses more dscriptive strings that can sometimes aid debugging

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      throw new RuntimeException("File " + filePath + " already exists");
+    }
+    file.deleteOnExit();
+    RandomAccessFile raf;
+    try {
+      raf = new RandomAccessFile(filePath, "rw");
+    } catch (FileNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+    long fileLen = Math.max(FILE_LENGTH, len);
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: Still reading through code...trying to understand how length is passed in.
We need upper bound on fileLen to 2G. PinotDataBuffer won't handle sizes larger than that.
Why is fileLen capped at 0.5G vs say 2G limit. It's better to have a large file than multiple small ones (truncate() on close() to limit size)

ALso, if it's 0.5G why not extend the file instead of mapping new file ?

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+  /**
+   * @param size size of memory to be mmaped
+   * @param columnName Name of the column for which memory is being allocated
+   * @return
+   */
+  @Override
+  public PinotDataBuffer allocate(long size, String columnName) {
+    addFileIfNecessary(size);
+    PinotDataBuffer buffer = _currentBuffer.view(_availableOffset, _availableOffset+size);
+    _availableOffset += size;
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: I think this _availableOffset calculation if off
First allocation: 512M, _availableOffset = 512M
Second allocation: 250M, _availableOffset = 500 + 250 = 750M (but we've allocated 512M buffer, _curFileLen = 512M), 
Third alloc: 100M,  then in addFileIfNecessary: len (100M) + _availableOffset(750M)  < _curFileLen is false  and we will allocate extra buffer

You probably should just = here instead of +=

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java [hunk 9]
```diff
@@ -108,18 +110,21 @@ public FixedByteSingleValueMultiColReader getReader() {
...
+   * @param columnName
    */
-  public FixedByteSingleColumnSingleValueReaderWriter(int numRowsPerChunk, int columnSizesInBytes, String description) {
+  public FixedByteSingleColumnSingleValueReaderWriter(int numRowsPerChunk, int columnSizesInBytes,
+      OffHeapMemoryManager memoryManager, String columnName) {
     chunkSizeInBytes = numRowsPerChunk * columnSizesInBytes;
     this.numRowsPerChunk = numRowsPerChunk;
     this.columnSizesInBytes = columnSizesInBytes;
-    this.description = description;
+    this.memoryManager = memoryManager;
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java]: Who has the responsibility to close() memoryManager  ? Please add to javadoc

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java (hunk 5 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: done, changed to RealtimeIndexOffHeapMemoryManager.

Kept the derived classes the same, though.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java (hunk 6 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: In the realtime indexing, this memory manager is created each time a segment is created. And then for each column, we may allocate memory multiple times. I will add these as comments in the base class.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java (hunk 2 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: agreed. done.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java (hunk 3 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: private does not work, so made it protected

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java (hunk 4 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/OffHeapMemoryManager.java]: the only higher layer here is realtime segment impl, and that logs (and swallows) the exception. IF we log there, we get the segment name  in the log, as it does for other things in the destroy() method.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java (hunk 9 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java]: The creator of memory manager is expected to close it (in our case there is only one creator, RealtimeSegmentImpl)

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/DirectMemoryManager.java (hunk 1 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/DirectMemoryManager.java]: yes, added.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java (hunk 7 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: length is decided by the caller who wants the buffer. All callers make sure that  length is less than 2g. they have their own algorithms for deciding the length (e.g. use prev len, or double prev len, etc.)

Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: I want to pick a number that is high enough so that we can get multiple buffers, but low enoug that we don't run out of disk trying to ingest 100 partitions in a host. 2g will be an overkill, so I picked 0.5g. I think I should reduce it a bit, maybe.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java (hunk 8 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: It cannot be = because we need to add the current buf size to get the next available offset. 

I ran a unit test with your test case, and am seeing the right results.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+   * @param dirPathName directory under which all mmap files are created.
+   * @param segmentName Name of the segment for which this memory manager allocates memory
+   */
+  public MmapMemoryManager(String dirPathName, String segmentName) {
+    super(segmentName);
+    _dirPathName = dirPathName;
+  }
+
+  private void addFileIfNecessary(long len) {
+    if (len + _availableOffset < _curFileLen) {
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: This condition should be <=, so that we can accommodate a chunk 1 byte longer in the same buffer. 

e.g, if a bufer is 4 bytes long, the first alloc asks for 1 byte. avail offset will become 1. At that point, we should be able to afford an allocation of 3 bytes in the same buffer

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java (hunk 8 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: In fact, we are a little conservative in the super-class, so I changed that, added a unit test for corner cases.

Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: Actually, _availableOffset is = to latest allocated size because addFileIfNecessary sets it to zero. So, += is effectively = size. That's different from your comment. Is this expected ?

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/RealtimeIndexOffHeapMemoryManager.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+  /**
+   * Method to be implemented by inheriting concrete classes
+   */
+  protected abstract void doClose();
+
+  /**
+   * Sub-classes are required to call this method after they successfully allocate a PinotDataBuffer.
+   * @param buffer
+   */
+  protected void onBufferAdded(PinotDataBuffer buffer) {
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/RealtimeIndexOffHeapMemoryManager.java]: I would have preferred allocate to be a wrapper for abstract method, say, allocateInternal instead of having all subclasses do right thing by calling onBufferAdded. Any extension can easily miss calling onBufferAdded and tests will rarely catch that

Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/RealtimeIndexOffHeapMemoryManager.java]: yes, agreed, will do

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java (hunk 8 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/writer/impl/MmapMemoryManager.java]: what if addIfNecessary does not create a new file?

File: pinot-core/src/test/java/com/linkedin/pinot/core/indexsegment/utils/OffHeapStringStoreTest.java [hunk 12]
```diff
@@ -19,14 +19,15 @@
...
+import com.linkedin.pinot.core.io.writer.impl.DirectMemoryManager;
 import com.linkedin.pinot.core.io.writer.impl.OffHeapStringStore;
 
 
 public class OffHeapStringStoreTest {
   private static Random RANDOM = new Random();
   @Test
   public void maxValueTest() throws Exception {
-    OffHeapStringStore store = new OffHeapStringStore();
+    OffHeapStringStore store = new OffHeapStringStore(new DirectMemoryManager("test"), "stringColumn");
```
Comment (mcvsubbu) [inline pinot-core/src/test/java/com/linkedin/pinot/core/indexsegment/utils/OffHeapStringStoreTest.java]: yes, agreed.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnMultiValueReaderWriter.java [hunk 13]
```diff
@@ -20,6 +20,7 @@
 import java.util.List;
 import com.linkedin.pinot.core.io.reader.impl.FixedByteSingleValueMultiColReader;
 import com.linkedin.pinot.core.io.readerwriter.BaseSingleColumnMultiValueReaderWriter;
+import com.linkedin.pinot.core.io.readerwriter.RealtimeIndexOffHeapMemoryManager;
```
Comment (rven1) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnMultiValueReaderWriter.java]: Should this class know about a realtime specific memory manager, or should we reference an interface? Wouldn't this class also be called for offline segments.

Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnMultiValueReaderWriter.java]: This class is not used by oflfine segments
